### ISSUE TITLE
[BUG] StackOverflowError when checking inductive invariant

### ISSUE THREAD

**Issue by will62794** [informalsystems/apalache#1336]
## Description

`StackOverflowError` is thrown by apalache when trying to check an inductive invariant for a basic TLA+ specification.

## Input specification

Attached folder contains TLA+ specification and config file used for repro. See [apalache-repro-22-02-13.zip](https://github.com/informalsystems/apalache/files/8057683/apalache-repro-22-02-13.zip).

## The command line parameters used to run the tool
```
apalache-v0.19.3/bin/apalache-mc check --max-error=150 --init=Safety --inv=Safety --length=1 learning_switch.tla
```
## Expected behavior

I expected apalache to produce a set of counterexamples to induction.

## Log files

See attached zip file ([apalache-repro-22-02-13.zip](https://github.com/informalsystems/apalache/files/8057683/apalache-repro-22-02-13.zip)).

## System information

- Apalache version: `apalache-v0.19.3` (build `52f9fe2`)
- OS: macOS 11.6.1, Apple M1 Silicon
- JDK version:
```
java version "16" 2021-03-16
Java(TM) SE Runtime Environment (build 16+36-2231)
Java HotSpot(TM) 64-Bit Server VM (build 16+36-2231, mixed mode, sharing)
```

## Additional context

It appears that this can be worked around by manually increasing the Java stack size i.e. running
```
JVM_ARGS="-Xss16M" apalache-v0.19.3/bin/apalache-mc check --max-error=150 --init=Safety --inv=Safety --length=1 learning_switch.tla
```
runs successfully and produces a set of counterexamples.

**Comment by konnov** [comment]
Hi @will62794! Thanks for the report. It seems that we have introduced a bottleneck in the recent versions, which introduce the experimental array encoding. We are investigating this.

In the meantime, you could use the older versions. For instance, [v0.17.4]( https://github.com/informalsystems/apalache/releases/tag/v0.17.4) works on your spec with default settings.

**Comment by konnov** [comment]
Closed via #1354
